File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

N, M = map(int, sys.stdin.readline().split())
trees = list(map(int,sys.stdin.readline().split()))

start, end = 1, max(trees)

# start가 end보다 커질때까지 반복
while start<=end:
    mid = (start+end)//2
    tree_sum = 0
    for tree in trees:
        if tree>mid:
            tree_sum+=tree-mid
    # 자른 나무의 합이 생각한거보다 크거나 같으면
    # 같은것도 넣어주는 이유 : start가 end보다 커져야 끝나고 end값을 출력할 것이라서
    #                    start=end일때 start를 올려줘야 나중에 출력값이 정상적으로 나옴
    if tree_sum>=M:
        start = mid+1
    else:
        end = mid-1


print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

/*
binary_search를 통해 어레이가 아닌 상수값을 찾는다.
하나가 아니기 때문에 가능한 경우에도 
low값을 하나씩 올려가면서 최대인 값을 계속 탐색한다.
*/

int N, M;

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    
    cin >> N >> M;
    vector<int> tree(N);
    for (int i=0; i<N; i++){
        cin >> tree[i];
    }
    int maximum = *max_element(tree.begin(), tree.end());

    int low = 1, high = maximum, ans = 0;

    while(low <= high){
        int mid = (low + high)/2;

        ll cnt = 0;  // 자른 랜선 개수 초기화
        for (int i=0; i<N; i++){
            cnt += max(0, tree[i]-mid);
        }

        // 가져가는 나무의 총 길이가 더 크거나 같으면
        if (cnt >= M){
            low = mid + 1;
            ans = max(ans, mid);
        }
        else {
            high = mid - 1;
        }
    }
    cout << ans;
    
    return 0;
}